### DISCUSSION TITLE
Ability to line-break toolbar at user-defined position

### DISCUSSION THREAD

**Discussion by mitchmeister** [chrisgurney/obsidian-note-toolbar#34] [Ideas]
First up!  Thanks so much.  Great contribution to the Obsidian community.

eg. For this note I would like to break the toolbar so that days of the week appear on the 2nd line.  Perhaps in addition to "URI", "Command" & "File" you could add "Line break"

<img width="855" alt="SCR-20240411-ncja" src="https://github.com/chrisgurney/obsidian-note-toolbar/assets/6250875/1dce8795-0d75-480e-82e9-a24c962d732b">

**Comment by chrisgurney** [comment]
@mitchmeister Thank you for the suggestion! (And I appreciate the screenshot!)

An enhancement I've thought about is the ability to create drop-downs / sub-toolbars, the advantage being it minimizes screen real-estate until those options are needed. e.g., Perhaps in your case, you would have a "Go To" item that opens a days of the week drop-down, or exposes a second toolbar below the first.

Might that work as a solution for your days of the week list?

**Reply by mitchmeister** [reply]
That would be great @chrisgurney !

**Reply by MrComputron** [reply]
I agree with this totally. It would br extremely useful to be able to organize the toolbar by row, particularly when on different devices it may render buttons to different rows in its present configuration. 

I think the nested Toolbar idea - or as I understand 
It the option to create a parent toolbar out of several child toolbars - would be a very good addition to the above concept as you could simply assign child toolbars to each row.

**Reply by calvin-1** [reply]
I just thought of this feature independently but how I imagined it could be implemented by linking between toolbars, if that is possible. that would add the ability to create nested or sub-toolbars. 

you could also create toolbars which act like a sub menu for commands - for example your main toolbar, "toolbar 1", could have a link to "toolbar 2" called "new note". toolbar 2 would have commands like "new note on new tab", "new from template" etc.. new from template could like to "toolbar 3" which has commands to insert certain templates and so on. 

it would essentially give you the ability to execute all commands and sub commands from the toolbar, or the floating button (which I think is superior on mobile).

**Reply by FelipeRearden** [reply]
@calvin-1  I think my FR #38  is related to your idea

**Comment by MrComputron** [comment]
Any changes on this item? This would turn Note Toolbar from a 10/10 to a 12/10.

I can't say enough how much functionality this plug-in adds to my day do day use, particularly on mobile, on par with Dataview for QOL feature.

**Reply by chrisgurney** [reply]
Once we get the beta sorted, sub-menus feel like a good candidate for the next release.

The way I think it could work is:

- For configuration you'd select "Menu" as a type for a toolbar item, and search for an existing toolbar.
- When using the toolbar, on clicking the item, it just opens a menu with the items from the selected toolbar.
  - From a toolbar menu (see the new beta) it just opens another menu in the same place.
- I'm not sure how well menus within menus will work -- will have to test!

**Reply by MrComputron** [reply]
I would go - if possible - with a "toolbar" item that when added has an option to be either "in-line" or "menu".

**In line** would basically insert the toolbar as a block made up of its buttons in their order
**Menu** would insert the toolbar as a drop down of the same toolbar.

I think both in line and drop down have their place and having the option to choose would be extremely useful.

I'm not sure if I missed a reply but are you thinking of enabling a "line break" item as originally posted?

**Reply by chrisgurney** [reply]
Thanks for the feedback.

> I'm not sure if I missed a reply but are you thinking of enabling a "line break" item as originally posted?

I'm not sold on this one, as various screen widths and device sizes don't guarantee a line break will be where you really want it to be. Opening a second row of buttons might work better conceptually, and that design pattern has been tried and tested in Microsoft's fluent design ribbon components, as an example.

As a step 1, I'll likely be adding menu support, as it works well on all screen sizes, and I've already done a lot of the foundational work (as it was needed for the floating buttons).

**Comment by chrisgurney** [comment]
@mitchmeister @MrComputron @calvin-1 Note Toolbar now supports both separators _and_ line breaks, in the latest beta!

Try out the release if you'd like, via BRAT, and let me know what you think. _It's recommended that you try this in another vault or backup your Note Toolbar's `data.json` first._

<img src="https://github.com/user-attachments/assets/13a36196-e121-4901-9599-8c94aaf8c6bf" alt="Example of a toolbar's settings with line breaks and separators added." width="600"/>

<img src="https://github.com/user-attachments/assets/35e08ded-accd-4763-a148-35b00ce05474" width="600" alt="Example toolbar with separators and line breaks."/>

**Comment by MrComputron** [comment]
I just tested out the new functionality and I like the separators for graphic control, a nice touch.
Unfortunately the line break items had no effect on the toolbar (on mobile).

**Reply by MrComputron** [reply]
Never mind my previous comment...I forgot to turn off "do not wrap" style on mobile.
Works perfectly!

This is really good work 👍

**Reply by chrisgurney** [reply]
Great, thank you very much for testing!

**Comment by chrisgurney** [comment]
Closing as this feature is now generally available.
